javascript/src/main/utils/BasicCourseTableHelpers.js

```diff
@@ -1,20 +1,52 @@
-
-export function reformatJSON(classes) {
+export function reformatJSON(classes,checks) {
   const sections = [];
+  var [cancelled,closed,full] = [false,false,false];
+  if(checks!==undefined) {
+    [cancelled,closed,full] = checks;
+  }
   classes.forEach(
     (course) => {
-      course.classSections.forEach(
-        (section) => {
-          section.course =
-          {
-          courseId: course.courseId,
-          title: course.title,
-          unitsFixed: course.unitsFixed
-          };
-         sections.push(section);
-       }
-      )
+      if( course.classSections[0].section.endsWith("0") && 
+      (((!cancelled) || (cancelled&&course.classSections[0].courseCancelled===null)) &&
+      ((!closed) || (closed&&course.classSections[0].classClosed===null)) &&
+      ((!full) || (full&&course.classSections[0].maxEnroll>=course.classSections[0].enrolledTotal)))) {
+        course.classSections.forEach(
+          (section) => {
+            if(((!cancelled) || (cancelled&&section.courseCancelled===null)) &&
+            ((!closed) || (closed&&section.classClosed===null)) &&
+            ((!full) || (full&&section.maxEnroll>=section.enrolledTotal))){
+              section.course =
+              {
+              courseId: course.courseId,
+              title: course.title,
+              unitsFixed: course.unitsFixed
+              };
+              sections.push(section);
+            } 
+          }
+        )
+      }
+      //For section tests
+      if(!course.classSections[0].section.endsWith("0")) {
+        course.classSections.forEach(
+          (section) => {
+            if(((!cancelled) || (cancelled&&section.courseCancelled===null)) &&
+            ((!closed) || (closed&&section.classClosed===null)) &&
+            ((!full) || (full&&section.maxEnroll>=section.enrolledTotal))){
+              section.course =
+              {
+              courseId: course.courseId,
+              title: course.title,
+              unitsFixed: course.unitsFixed
+              };
+              sections.push(section);
+            } 
+          }
+        )
+      }
+      
```

Comment:
This logic needs to be refactored into several much smaller, more well contained functions.

It is WAY too big and complicated.

I realize that the task is big and complicated, but that doesn't mean the code has to be.

Reply:
Start with these lines:
```
if(((!cancelled) || (cancelled&&section.courseCancelled===null)) &&
            ((!closed) || (closed&&section.classClosed===null)) &&
            ((!full) || (full&&section.maxEnroll>=section.enrolledTotal)))
```

Factor that into a function.  You use it twice, and it's ugly code.  That's TWO reasons why it should definitely be factored out.

Put the three variables `cancelled`, `closed` and `full` into a single parameter called filters.

So the function will be something like:

```
const sectionSelected  = (section, filters) => (
    ((!filters.cancelled) || ( filters.cancelled&&section.courseCancelled===null)) &&
    ((!filters.closed) || (filters.closed&&section.classClosed===null)) &&
     ((!filters.full) || (filters.full&&section.maxEnroll>=section.enrolledTotal)) 
 );
```

Then factor out

```
 section.course =
              {
              courseId: course.courseId,
              title: course.title,
              unitsFixed: course.unitsFixed
              };
```

Into something like

```
const addCourse = (section) => {
      section.course =
              {
              courseId: course.courseId,
              title: course.title,
              unitsFixed: course.unitsFixed
              };
}
```


Then this whole 30-40 lines of code will fold down to about 10.

Reply:
And, it will be WAY easier to test.  You'll be able to unit test the smaller functions separately instead of ONLY being able to test them in the context of complicated boolean logic.  That's super hard.